### DISCUSSION TITLE
Open file in EDITOR

### DISCUSSION THREAD

**Discussion by 037g** [PatrickF1/fzf.fish#104] [Archive] [answered]
Hi,

Thank you so much for making this.

Previously, I could use CTRL-O and it would let me open a file in the system editor. This does not seem to be available anymore. Can you please help direct me to the best way to accomplish this?

When I add this `bind --mode insert \co '__fzf_open --editor'` in my `conf.d/fzf.fish`
I get: `fish: Unknown command: __fzf_open`

Thank You

**Comment by PatrickF1** [comment]
I didn't implement that. The closest thing is to use type `vim` or whatever you use and then hit Ctrl+F. I'm not going to implement a tool just for opening files because I want the commands to be composable. Sorry about that.

**Comment by kidonng** [comment]
@PatrickF1 Not sure if you are interested, do you think it is favorable to output everything to stdout when executing a command substitution?

Currently all functions insert the output into the command line, which is fine for interactive usage, but not very interpolatable.

We can rewrite everything into something like:

```fish
if status is-command-substitution
    printf $output
else
    commandline --current-token --replace $output
end
```

That way people can reuse existing functions to achieve more niche usage, for example this feature request can become `$EDITOR (__fzf_search_current_dir)`.

**Reply by PatrickF1** [reply]
Oof this is a tough one. Am kind of split. Would love to hear more feedback from people before doing it. Let me get back to you.

**Reply by PatrickF1** [reply]
@IlanCosman said:
> Could you explain the use case for running this plugin non-interactively? I don't quite understand the motivation.

I think the idea is to make fzf.fish usable as just a library of functions for users to use to create their own key bindings.
Then people can do
`bind [key] vim (__fzf_search_current_dir)` or `bind [key] git add (__fzf_search_git_status)`

**Reply by IlanCosman** [reply]
Ah, okay, yah I like that idea. 👍

**Reply by PatrickF1** [reply]
Ok I've decided it's a good addition and we can shift in the direction of encouraging users to build on top of the plugin and create custom key bindings for all their specific use cases. However, I'd really dislike that even when in command substitution, a lot of interactive mode code runs. Is that a technical challenge we can overcome?

**Reply by IlanCosman** [reply]
I'm not sure I understand the question. These are command substitutions in name only. For your examples like `bind [key] git add (__fzf_search_git_status)`, the user would still be using the plugin interactively. All that needs to change is the output method, which is what Kid's PR does. Could you explain?

**Reply by PatrickF1** [reply]
All of the stuff scanning current token and doing stuff with it would be irrelevant. Maybe I'm making too big of a deal out of it

**Reply by ovv** [reply]
I think we could split the functions in two:
1. A base you can use in substitution (taking a token as optional parameter)
2. A wrapper of 1, that deals with what's necessary to make it run with a key binding and fetch the token

**Reply by PatrickF1** [reply]
Yeah that's what I'm thinking too. Gonna take a lot of work though!

**Reply by ovv** [reply]
my 0,02$, we might want to go ahead with #110 in the meantime.

Personally I would love to do something like `abbr -a ga 'git add (__fzf_search_git_status)'`

**Reply by tommyknows** [reply]
Is there a solution for doing `bind [key] vim (_fzf_search_current_dir)`? Not quite sure what changed in the meantime, but I'd definitely also be interested in such a feature :-)

**Comment by PatrickF1** [comment, marked as answer]
Hi all, opening a file in editor is now very possible with a bit of work. With the merging of https://github.com/PatrickF1/fzf.fish/pull/140, it is now possible to configure a key binding to open the current line in vim. This has a working example: https://github.com/junegunn/fzf/issues/1360. Just combine it with `fzf_dir_opts`.

**Comment by kidonng** [comment]
I hope this doesn't mean you will ditch #110 because it's from a different aspect.

**Reply by PatrickF1** [reply]
No I even have an idea of how to make it work but it's not high on my priority list currently. I will open a new discussion to track this idea.
EDIT: here is the new conversation. please direct your comments there https://github.com/PatrickF1/fzf.fish/discussions/144

**Comment by notDavid** [comment]
Hi there, thank you for making this!

I’m trying to get this to work, but i’m not sure what i’m doing wrong…  here’s what i did:

`set --export fzf_dir_opts '--bind \"ctrl-v:execute(vim {} < /dev/tty)\"'`

However, when attempting to start fzf.fish with `Ctrl+Alt+F` i get this error:

`unknown option: --bind \"ctrl-v:execute (vim {7 < /dev/tty)\" `

Could you please advise? thank you!

```
fish, version 3.3.1
fzf 0.28.0 (brew)
```

PS.
executing `fzf --bind 'ctrl-v:execute(vim {} < /dev/tty)'` on the commandline without using `fzf.fish` does work fine.

**Reply by PatrickF1** [reply]
Hi David, can you see if the discussion here helps? https://github.com/PatrickF1/fzf.fish/issues/203

**Reply by notDavid** [reply]
Hi Patrick, thank you for your reply! I’m not sure how to translate this solution into my situation. I tried this, and the problem remains the same:

`set --export fzf_dir_opts '--bind "ctrl-v:execute(vim {} &> /dev/tty)”'`

However, executing `set v (ls | fzf --bind "ctrl-v:execute(vim {} &> /dev/tty)")` directly in the fish shell without using `fzf.fish` works fine.

Is what i did correct, or is the value i’m setting `fzf_dir_opts` to wrong?

**Reply by PatrickF1** [reply]
how about `set --export fzf_dir_opts --bind "ctrl-v:execute(vim {} &> /dev/tty)`
basically remove the single quotes wrapping the whole thing

**Reply by notDavid** [reply]
Thank you so much @PatrickF1, this works! 👌🏼

`set --export fzf_dir_opts --bind "ctrl-v:execute(vim {} &> /dev/tty)"`
